packages/contracts/test/exchange/internal.ts

```diff
@@ -0,0 +1,326 @@
+}
+
+async function _getOverflowErrorForSendTransaction(): Promise<Error> {
+    const errMsg = await getGanacheOrGethError('invalid opcode', 'always failing transaction');
+    return new Error(errMsg);
+}
+
+async function referenceGetPartialAmountAsync(
+    numerator: BigNumber,
+    denominator: BigNumber,
+    target: BigNumber,
+): Promise<BigNumber> {
+    const overflowError = await _getOverflowErrorForCall();
+    if (numerator.greaterThan(MAX_UINT256)) {
+        throw overflowError;
+    } else if (denominator.greaterThan(MAX_UINT256)) {
+        throw overflowError;
+    } else if (denominator.eq(new BigNumber(0))) {
+        throw overflowError;
+    } else if (target.greaterThan(MAX_UINT256)) {
+        throw overflowError;
+    }
+    const product = numerator.mul(target);
+    if (product.greaterThan(MAX_UINT256)) {
+        throw overflowError;
+    }
+    return product.dividedToIntegerBy(denominator);
+}
+
+describe('Exchange core internal functions', () => {
+    let testExchange: TestExchangeInternalsContract;
+    let overflowErrorForCall: Error | undefined;
+    let overflowErrorForSendTransaction: Error | undefined;
+
+    before(async () => {
+        await blockchainLifecycle.startAsync();
+    });
+    after(async () => {
+        await blockchainLifecycle.revertAsync();
+    });
+    before(async () => {
+        testExchange = await TestExchangeInternalsContract.deployFrom0xArtifactAsync(
+            artifacts.TestExchangeInternals,
+            provider,
+            txDefaults,
+        );
+        overflowErrorForCall = await _getOverflowErrorForCall();
+        overflowErrorForSendTransaction = await _getOverflowErrorForSendTransaction();
+    });
+    // Note(albrow): Don't forget to add beforeEach and afterEach calls to reset
+    // the blockchain state for any tests which modify it!
+
+    describe('addFillResults', async () => {
+        function makeFillResults(value: BigNumber): FillResults {
+            return {
+                makerAssetFilledAmount: value,
+                takerAssetFilledAmount: value,
+                makerFeePaid: value,
+                takerFeePaid: value,
+            };
+        }
+        async function referenceAddFillResultsAsync(
+            totalValue: BigNumber,
+            singleValue: BigNumber,
+        ): Promise<FillResults> {
+            // Note(albrow): Here, each of totalFillResults and
+            // singleFillResults will consist of fields with the same values.
+            // This should be safe because none of the fields in a given
+            // FillResults are ever used together in a mathemetical operation.
+            // They are only used with the corresponding feild from *the other*
+            // FillResults, which are different.
+            const totalFillResults = makeFillResults(totalValue);
+            const singleFillResults = makeFillResults(singleValue);
+            // Note(albrow): _.mergeWith mutates the first argument! To
```

Comment:
Let's stick to the convention of prefixing comments about hacks/workarounds with `// HACK`.